crypto/evp/p_lib.c

```diff
@@ -539,8 +540,16 @@ EC_KEY *EVP_PKEY_get1_EC_KEY(EVP_PKEY *pkey)
 
 int EVP_PKEY_set1_DH(EVP_PKEY *pkey, DH *key)
 {
-    int type = DH_get0_q(key) == NULL ? EVP_PKEY_DH : EVP_PKEY_DHX;
-    int ret = EVP_PKEY_assign(pkey, type, key);
+    int type = EVP_PKEY_DH;
+    int ret;
+
+    /*
+     * Use DHX for any DH key that has a q parameter,
+     * except for named groups (for backwards compatibility reasons).
+     */
+    if (DH_get0_q(key) != NULL && dh_get0_nid(key) == NID_undef)
+        type = EVP_PKEY_DHX;
```

Comment:
I'm not sure this is the right answer - and wasn't the point of my comments. It seems to me right that the type should be DHX if a "q" is present. The code in libssl I'm concerned about is only in master so we need not be worried about backwards compat there. The test I was worried about, was only an issue because it changed the intent of the test, i.e. the point was that if you create an EVP_PKEY from a DH without a q then you you should get EVP_PKEY_DH. Fudging things to force that to happen in libcrypto seems like the wrong answer. Instead the test should be changed to use a different DH group that doesn't have a q (perhaps a non-named group if needs be).

Reply:
I have changed the test input so it has no q and this produces a dhx_type. Got rid of the extra check for a named group when determining the type, so all named groups use the DHX type.

Reply:
@mattcaswell can you review please/